rcl_yaml_param_parser/src/parser.c

```diff
@@ -0,0 +1,1362 @@
+  if (NULL == params_st->params) {
+    rcl_yaml_node_struct_fini(params_st, allocator);
+    RCUTILS_SAFE_FWRITE_TO_STDERR("Error allocating mem");
+    return NULL;
+  }
+
+  params_st->num_nodes = 0U;
+
+  return params_st;
+}
+
+///
+/// Free param structure
+/// NOTE: If there is an error, would recommend just to safely exit the process instead
+/// of calling this free function and continuing
+///
+void rcl_yaml_node_struct_fini(
+  rcl_params_t * params_st,
+  const rcl_allocator_t allocator)
+{
+  uint32_t node_idx;
+  uint32_t parameter_idx = 0U;
+
+  if (NULL == params_st) {
+    return;
+  }
+
+  for (node_idx = 0; node_idx < params_st->num_nodes; node_idx++) {
+    char * node_name = params_st->node_names[node_idx];
+    if (NULL != node_name) {
+      allocator.deallocate(node_name, NULL);
+    }
+
+    if (NULL != params_st->params) {
+      char * param_name;
+      rcl_variant_t * param_var;
+      for (parameter_idx = 0; parameter_idx < params_st->params[node_idx].num_params;
+        parameter_idx++)
+      {
+        if ((NULL != params_st->params[node_idx].parameter_names) &&
+          (NULL != params_st->params[node_idx].parameter_values))
+        {
+          param_name = params_st->params[node_idx].parameter_names[parameter_idx];
+          param_var = &(params_st->params[node_idx].parameter_values[parameter_idx]);
+          if (NULL != param_name) {
+            allocator.deallocate(param_name, NULL);
+          }
+
+          if (NULL != param_var) {
+            if (NULL != param_var->bool_value) {
+              allocator.deallocate(param_var->bool_value, NULL);
+            } else if (NULL != param_var->integer_value) {
+              allocator.deallocate(param_var->integer_value, NULL);
+            } else if (NULL != param_var->double_value) {
+              allocator.deallocate(param_var->double_value, NULL);
+            } else if (NULL != param_var->string_value) {
+              allocator.deallocate(param_var->string_value, NULL);
+            } else if (NULL != param_var->bool_array_value) {
+              if (NULL != param_var->bool_array_value->values) {
+                allocator.deallocate(param_var->bool_array_value->values, NULL);
+              }
+              allocator.deallocate(param_var->bool_array_value, NULL);
+            } else if (NULL != param_var->integer_array_value) {
+              if (NULL != param_var->integer_array_value->values) {
+                allocator.deallocate(param_var->integer_array_value->values, NULL);
+              }
+              allocator.deallocate(param_var->integer_array_value, NULL);
+            } else if (NULL != param_var->double_array_value) {
+              if (NULL != param_var->double_array_value->values) {
+                allocator.deallocate(param_var->double_array_value->values, NULL);
+              }
+              allocator.deallocate(param_var->double_array_value, NULL);
+            } else if (NULL != param_var->string_array_value) {
```

Comment:
Missing dealloation of `param_var->string_array_value->data`. Recommend using `rcutils_string_array_fini()` instead.

Reply:
337a755